receiver/dotnetdiagnosticsreceiver/dotnet/request.go

```diff
@@ -0,0 +1,185 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package dotnet
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+	"io"
+	"math"
+	"sort"
+	"strconv"
+	"strings"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/receiver/dotnetdiagnosticsreceiver/network"
+)
+
+// RequestWriter submits the initial request to the dotnet diagnostics backend,
+// after which it starts sending us metrics at the specified interval.
+// ByteBuffer is swappable for testing.
+// For docs on the protocol, see
+// https://github.com/dotnet/diagnostics/blob/master/documentation/design-docs/ipc-protocol.md
+type RequestWriter struct {
+	w           io.Writer
+	intervalSec int
+	// providerNames (aka event sources) indicate which counter groups to get metrics for. e.g. "System.Runtime"
+	providerNames []string
+}
+
+func NewRequestWriter(w io.Writer, intervalSec int, providerNames ...string) *RequestWriter {
+	return &RequestWriter{w: w, intervalSec: intervalSec, providerNames: providerNames}
+}
+
+func (w *RequestWriter) SendRequest() error {
+	req := w.createRequest()
+	_, err := w.w.Write(req)
+	return err
+}
+
+// The following constants come from:
+// https://github.com/dotnet/diagnostics/blob/master/src/Microsoft.Diagnostics.NETCore.Client/DiagnosticsIpc/IpcCommands.cs
+
+// Corresponds to DiagnosticsServerCommandSet.EventPipe
+const eventPipeCommand = 2
+
+// Corresponds to EventPipeCommandId.CollectTracing2
+const collectTracing2CommandID = 3
+
+func (w *RequestWriter) createRequest() []byte {
+	cfgReq := newConfigRequest(w.intervalSec, w.providerNames...)
+	payload := cfgReq.serialize()
+	hdr := &requestHeader{
+		commandSet: eventPipeCommand,
+		commandID:  collectTracing2CommandID,
+	}
+	hdrBytes := hdr.serialize(len(payload))
+	return append(hdrBytes, payload...)
+}
+
+// requestHeader
+type requestHeader struct {
+	commandSet byte
+	commandID  byte
+}
+
+const magic = "DOTNET_IPC_V1"
+const requestHeaderSize = 20
+
+func (h requestHeader) serialize(payloadSize int) []byte {
+	buf := &bytes.Buffer{}
+	_, _ = buf.Write([]byte(magic + "\000"))
+	totSize := uint16(requestHeaderSize + payloadSize)
+	_ = binary.Write(buf, network.ByteOrder, totSize)
+	_, _ = buf.Write([]byte{h.commandSet, h.commandID, 0, 0})
```

Comment:
A little curious why these errors aren't handled (though unlikely to occur).  Is it related to coverage requirements and difficulty mocking?

Reply:
Good question -- I actually handled them and they were covered/mocked in a previous commit, but I removed the error handling and related tests because bytes.Buffer#Write always returns a nil error. I'll add a comment to make this clear.